src/twisted/internet/defer.py

```diff
@@ -161,6 +172,151 @@ def maybeDeferred(f, *args, **kw):
 
 
 
+def extractCurrentResult(deferred, raises=True):
+    """
+    Extract the current result from a L{Deferred} that has already fired all of
+    its callbacks.
+
+    Take a deferred that has already fired all of its callbacks and extract and
+    return the result of the L{Deferred}. This process of extracting the result
+    will consume the result, error or otherwise. If the L{Deferred} has an
+    error then the process of extracting the result will, by default, re-raise
+    the original exception unless the L{raises} parameter is set to L{False}.
+
+    The primary purpose of this function is to allow code to be written that is
+    internally asynchronous but which can be consumed without an event loop in
+    a synchronous fashion by using L{extractCurrentResult} at the boundaries.
+    This enables libraries to be written that can function both with Twisted
+    and with synchronous code instead of requiring an asynchronous and a
+    synchronous implementation for all functionality.
+
+    @attention: This function CANNOT take a Deferred with pending asynchronous
+        work to be done and make it synchronous. It can ONLY extract results
+        from a L{Deferred} which has already fired its callbacks.
+
+    @type deferred: Deferred
+    @param deferred: The Deferred from which results will be extracted..
+
+    @type raises: bool
+    @param raises: Whether or not to raise the original exception (L{True}) or
+        to return the failure object (L{False}).
+
+    @raise NoCurrentResult: If C{deferred} has pending work to be done
+        and has no current result to extract.
+
+    @return: The result of the L{Deferred}.
+    """
+    def done(result):
+        done.result = result
+
+    # We need a place to store the result, and a way to determine the
+    # difference between no result and any other possible object. So we'll
+    # create a sigil object and stash it locally and default our result to
+    # that.
+    # Note: This *has* to happen before adding the done callback to the
+    #       Deferred or else the callback can fire prior to setting our
+    #       default, thus causing the default to override the value from the
+    #       callback.
+    nothing = done.result = object()
```

Comment:
Maybe having just one of these (and stashing it as e.g. `NoCurrentResult._SENTINEL` or whatever) is a bit shorter than creating it each time?